c/common/src/codingstandards/c/OutOfBounds.qll

```diff
@@ -0,0 +1,772 @@
+
+  predicate isBufferSizeExprGreaterThanSourceSizeExpr(
+    Expr bufferUse, Expr bufferSize, Expr sizeSource, PointerToObjectSource sourceBufferAllocation,
+    int bufSize, int size, BufferAccessLibraryFunctionCall fc, int offset, Expr base
+  ) {
+    exists(float sizeMult |
+      (
+        bufferUse = fc.getWriteArg() and
+        bufferSize = fc.getWriteSizeArg() and
+        sizeMult = fc.getWriteSizeArgMult()
+        or
+        bufferUse = fc.getReadArg() and
+        bufferSize = fc.getReadSizeArg() and
+        sizeMult = fc.getReadSizeArgMult()
+      ) and
+      (
+        offset = 0 and
+        base = bufferSize and
+        bufferUseComputableBufferSize(bufferUse, sourceBufferAllocation, bufSize) and
+        sizeExprComputableSize(bufferSize, sizeSource, size) and
+        (
+          bufSize - getArithmeticOffsetValue(bufferUse) <
+            (sizeMult * (size + getArithmeticOffsetValue(bufferSize)).(float))
+          or
+          size = 0
+        )
+      )
+    )
+  }
+
+  predicate isBufferSizeOffsetOfGVN(
+    BufferAccessLibraryFunctionCall fc, Expr bufferSize, Expr bufferUse,
+    DynamicAllocationSource source, Expr sourceSizeExpr, Expr sourceSizeExprBase,
+    int sourceSizeExprOffset, int sizeMult, int sizeArgOffset, int bufferArgOffset
+  ) {
+    (
+      bufferUse = fc.getWriteArg() and
+      bufferSize = fc.getWriteSizeArg() and
+      sizeMult = fc.getWriteSizeArgMult()
+      or
+      bufferUse = fc.getReadArg() and
+      bufferSize = fc.getReadSizeArg() and
+      sizeMult = fc.getReadSizeArgMult()
+    ) and
+    sourceSizeExpr = source.getSizeExprSource(sourceSizeExprBase, sourceSizeExprOffset) and
+    bufferUseNonComputableSize(bufferUse, source) and
+    not globalValueNumber(sourceSizeExpr) = globalValueNumber(bufferSize) and
+    exists(Expr offsetExpr |
+      offsetExpr = bufferSize.getAChild*() and
+      sizeArgOffset = getArithmeticOffsetValue(offsetExpr)
+    ) and
+    bufferArgOffset = getArithmeticOffsetValue(bufferUse) and
+    sourceSizeExprOffset + bufferArgOffset < sizeArgOffset
+  }
+
+  predicate problems(BufferAccessLibraryFunctionCall fc, string msg) {
+    exists(Expr bufferUse, PointerToObjectSource source |
+      exists(int bufSize, int size, Expr bufferSize, Expr sizeSource |
+        isBufferSizeExprGreaterThanSourceSizeExpr(bufferUse, bufferSize, sizeSource, source,
+          bufSize, size, fc, _, _) and
+        msg = "Buffer size is smaller than size arg."
+      )
+      or
+      exists(int i |
+        fc.getTarget().(BufferAccessLibraryFunction).getANullTerminatedParameterIndex(i) and
+        fc.getArgument(i) = bufferUse and
+        source.isNotNullTerminated() and
+        hasFlowFromBufferOrSizeExprToUse(source, bufferUse.getAChild*()) and
+        msg = "Buffer " + bufferUse.toString() + " is not null-terminated."
+      )
+      or
+      isBufferSizeOffsetOfGVN(fc, _, bufferUse, source, _, _, _, _, _, _) and
+      msg = "Buffer size is offset of GVN."
```

Comment:
This alert message needs improving (ideally with the actual offset numbers if possible).